### ISSUE TITLE
Question on tf.control_dependencies(update_ops) in ModelDesc.build_graph()

### ISSUE THREAD

**Issue by EchoWho** [tensorpack/tensorpack#1087]
Suppose there are some non-gradient updates such as the following:

https://github.com/tensorpack/tensorpack/blob/02c38f26d938cf6cdad4a09ada754ed950d024d7/examples/PennTreebank/PTB-LSTM.py#L96-L97 

The update ops modify the inputs (init state variable). The  tf.control_dependencies run these updates before cost is computed, which is before the gradient is computed. 

**Will this update affect the gradient?** In this case, the gradient and its subsequent trainer.train_op are done after the update ops. 
We can not simply put such updates in UpdateOperation collections either, because Tensorflow does not guarantee the order of independent operations. (To see that the independent ops can be run in strange order, see the example in the end.) 

In general, what's your recommendation on implementing hooks that are run after trainer.train_op in tensorpack, since the trainer.train_op is constructed after the callbacks are specified in the training config. (The example in the end also shows how to do this in general TF). 

It seems to me Tensorpack can only do so is somewhat as follows. First create a ModelDesc method for such update ops, then create a Callback whose _setup_graph uses self.trainer.train_ops as the control dependency to call the abovementioned method via an input that's like (lambda : model.xxxx()). 
Furthermore, during testing, a new callback is needed, because now there is no train_op, but instead we have a list of output tensors that are required as dependencies. 

Are there cleaner ways to do this (i.e., a cleaner way to specify the dependencies for call backs

Example that shows the strange ordering in tensorflow, and how to fix it for hooks. 
```
import tensorflow as tf

class AddOneHook(tf.train.SessionRunHook):
    
    def __init__(self, x, dependencies=[]):
        self.x = x
        self.dependencies = dependencies

    def begin(self):
        with tf.control_dependencies(self.dependencies):
            self.op = x.assign_add(1)

    def before_run(self, run_context):
        return tf.train.SessionRunArgs(self.op, None, None)

if __name__ == '__main__':
    x = tf.get_variable('x', [], initializer=tf.constant_initializer(0))
    x_times_2 = x * 2

    print('-' * 79)
    hook = AddOneHook(x, [x_times_2])
    with tf.train.MonitoredSession(session_creator=tf.train.ChiefSessionCreator(), hooks=[hook]) as sess:
        for _ in range(4):
            print(sess.run([x_times_2]))

    print('-' * 79)
    hook = AddOneHook(x, [])
    with tf.train.MonitoredSession(session_creator=tf.train.ChiefSessionCreator(), hooks=[hook]) as sess:
        for _ in range(4):
            print(sess.run([x_times_2]))

    """
    Outputs:
    --------------------
    [0.0]
    [2.0]
    [4.0]
    [6.0]
    --------------------
    [2.0]
    [4.0]
    [6.0]
    [8.0]
% Note in the second round there is no dependency control and the assign_add happens before the mul. 
    """
```

**Comment by ppwwyyxx** [comment]
> In general, what's your recommendation on implementing hooks that are run after trainer.train_op in tensorpack, since the trainer.train_op is constructed after the callbacks are specified in the training config. (The example in the end also shows how to do this in general TF).

[launch_train_with_config](https://tensorpack.readthedocs.io/modules/train.html#tensorpack.train.launch_train_with_config) is almost equivalent to the following two lines of code:
```
    trainer.setup_graph(
        model.get_inputs_desc(), input,
        model.build_graph, model.get_optimizer)
    trainer.train_with_defaults(
        callbacks=config.callbacks,
        monitors=config.monitors,
        session_creator=config.session_creator,
        session_init=config.session_init,
        steps_per_epoch=config.steps_per_epoch,
        starting_epoch=config.starting_epoch,
        max_epoch=config.max_epoch,
        extra_callbacks=config.extra_callbacks)
```
You can run the first line first, which will create the train_op in trainer.
You can then construct your callbacks, and call the second line to start training.

**Comment by ppwwyyxx** [comment]
Feel free to reopen if the above does not answer your questions.
